### DISCUSSION TITLE
Hydration Issue for Conditional Rendering

### DISCUSSION THREAD

**Discussion by edward1127** [vercel/next.js#58846] [Help]
### Summary

I have a hook that returns the client's screen size. 
```
export const useWindowDimensions = () => {
  const [dimesion, setDimension] = useState(getWindowDimensions());

  const isXs = useMemo(() => dimesion.width < 576, [dimesion]);
  const isSm = useMemo(() => dimesion.width < 768, [dimesion]);
  const isMd = useMemo(() => dimesion.width < 992, [dimesion]);
  const isLg = useMemo(() => dimesion.width < 1200, [dimesion]);
  const isXl = useMemo(() => dimesion.width >= 1200, [dimesion]);

  useEffect(() => {
    const handleResize = () => {
      setDimension(getWindowDimensions);
    };
    window.addEventListener("resize", handleResize);
    return () => window.removeEventListener("resize", handleResize);
  }, []);

  return { isXs, isSm, isMd, isLg, isXl, dimesion };
};

```

If I use isSm to conditionally render between components A and B, it will throw a hydration error. This is because on the server side, isSm will be true, but on the client side, it will be false. Does anyone know how I can resolve this issue?

**Comment by icyJoseph** [comment]
Well, I guess `getWindowDimensions` is returning conditionally for client and server side. 

You can:

0. Use CSS
1. Set the initial state of `dimension` to a sensible default, for example, `null`, or an empty object, or an object with defaults `{ width: 500, height: 500}`
2. Or, put the usage of this hook behind a client side activated flag, for example in this gist, Option 2, puts `useLayoutEffect` behind a client side activated flag ~ That'll impact SEO though

- https://gist.github.com/gaearon/e7d97cdf38a2907924ea12e4ebdf3c85

## Edit

Forgot to share the gist link

**Reply by edward1127** [reply]
thanks for replying! 
0.  In my case, component A and component B are two different components and don't share many things in common, I don't think CSS can achieve that. 
1. if I set the default to { width: 500, height: 500} then in a big screen, since now both server and client's isSm is false, it appears that the problem is resolved, but when you open it will the actual small screen, the hydration issue is still there. 
2. If you mean that I can use `typeof window !== "undefined"`, it's actually the same as 1 since when you change to other screen size, it will throw hydration error.

**Reply by edward1127** [reply]
it seems that the only way to solve it is to disable ssr?

**Reply by icyJoseph** [reply]
Well, yeah, the server has no knowledge of client dimensions and such, so you kind of have to find a balance here... If the cost of having various groups of HTML (hiding them using CSS) is too large, and if a sensible fallback is not possible, then yeah, you'd need to have a client side activated flag.

> If you mean that I can use typeof window !== "undefined", it's actually the same as 1 since when you change to other screen size, it will throw hydration error.

No I don't mean that, I actually suppose that's what you have in you `getWindowDimensions`, which is my very first statement. Rendering using that condition is just asking for trouble.

**Reply by icyJoseph** [reply]
I realize I did not share the gist, with the client side activated example...

- https://gist.github.com/gaearon/e7d97cdf38a2907924ea12e4ebdf3c85
- You could also use `next/dynamic` with ssr set to false I guess

**Reply by edward1127** [reply]
yep, I think setting ssr to false will probably be easier in my case, thanks for the suggestions!

**Reply by GabenGar** [reply]
If your component renders at two different paths to the point it causes hydration issues, then it's time to put it behind a loading spinner on server and only resolve on the client.

**Reply by edward1127** [reply]
thanks for sharing your solution
